File: pomodoro.go
package main

import (
	"encoding/json"
	"log"
	"net/http"
	"time"
)

// ==========================================
// POMODORO TIMER HANDLER
// ==========================================

// handlePomodoro manages the Pomodoro timer state and settings
func handlePomodoro(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")

	if r.Method == http.MethodGet {
		// Return current Pomodoro state and settings
		mutex.Lock()
		response := map[string]interface{}{
			"session":  pomodoroSession,
			"settings": pomodoroSettings,
		}
		mutex.Unlock()
		json.NewEncoder(w).Encode(response)
		return
	}

	if r.Method == http.MethodPost {
		var req struct {
			Action string `json:"action"` // "start", "pause", "resume", "reset", "skip"
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "Invalid request body", http.StatusBadRequest)
			return
		}

		mutex.Lock()
		switch req.Action {
		case "start":
			pomodoroSession.Active = true
			pomodoroSession.IsPaused = false
			pomodoroSession.StartedAt = time.Now()
			pomodoroSession.Mode = "work"
			pomodoroSession.TimeRemaining = pomodoroSettings.WorkDuration
			log.Printf("🍅 Pomodoro started: %d min work session", pomodoroSettings.WorkDuration/60)

		case "pause":
			if pomodoroSession.Active && !pomodoroSession.IsPaused {
				pomodoroSession.IsPaused = true
				pomodoroSession.PausedRemaining = pomodoroSession.TimeRemaining
				log.Printf("🍅 Pomodoro paused: %d seconds remaining", pomodoroSession.TimeRemaining)
			}

		case "resume":
			if pomodoroSession.Active && pomodoroSession.IsPaused {
				pomodoroSession.IsPaused = false
				pomodoroSession.StartedAt = time.Now()
				log.Printf("🍅 Pomodoro resumed: %d seconds remaining", pomodoroSession.TimeRemaining)
			}

		case "reset":
			pomodoroSession.Active = false
			pomodoroSession.IsPaused = false
			pomodoroSession.Mode = "work"
			pomodoroSession.TimeRemaining = pomodoroSettings.WorkDuration
			pomodoroSession.CyclesCompleted = 0
			log.Printf("🍅 Pomodoro reset")

		case "skip":
			// Skip to next phase
			if pomodoroSession.Mode == "work" {
				pomodoroSession.CyclesCompleted++
				if pomodoroSession.CyclesCompleted >= pomodoroSettings.CyclesUntilLong {
					pomodoroSession.Mode = "longBreak"
					pomodoroSession.TimeRemaining = pomodoroSettings.LongBreak
					pomodoroSession.CyclesCompleted = 0
					log.Printf("🍅 Pomodoro: Long break started (%d min)", pomodoroSettings.LongBreak/60)
				} else {
					pomodoroSession.Mode = "break"
					pomodoroSession.TimeRemaining = pomodoroSettings.BreakDuration
					log.Printf("🍅 Pomodoro: Break started (%d min), cycle %d/%d",
						pomodoroSettings.BreakDuration/60, pomodoroSession.CyclesCompleted, pomodoroSettings.CyclesUntilLong)
				}
			} else {
				pomodoroSession.Mode = "work"
				pomodoroSession.TimeRemaining = pomodoroSettings.WorkDuration
				log.Printf("🍅 Pomodoro: Work session started (%d min)", pomodoroSettings.WorkDuration/60)
			}
			pomodoroSession.StartedAt = time.Now()
			pomodoroSession.IsPaused = false

		default:
			mutex.Unlock()
			jsonError(w, "Invalid action: "+req.Action, http.StatusBadRequest)
			return
		}
		response := map[string]interface{}{
			"session":  pomodoroSession,
			"settings": pomodoroSettings,
		}
		mutex.Unlock()
		json.NewEncoder(w).Encode(response)
		return
	}

	if r.Method == http.MethodPut {
		// Update Pomodoro settings
		var req struct {
			WorkDuration    *int  `json:"workDuration"`
			BreakDuration   *int  `json:"breakDuration"`
			LongBreak       *int  `json:"longBreak"`
			CyclesUntilLong *int  `json:"cyclesUntilLong"`
			ShowInCycle     *bool `json:"showInCycle"`
		}
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "Invalid request body", http.StatusBadRequest)
			return
		}

		mutex.Lock()
		if req.WorkDuration != nil && *req.WorkDuration >= 60 && *req.WorkDuration <= 3600 {
			pomodoroSettings.WorkDuration = *req.WorkDuration
		}
		if req.BreakDuration != nil && *req.BreakDuration >= 60 && *req.BreakDuration <= 1800 {
			pomodoroSettings.BreakDuration = *req.BreakDuration
		}
		if req.LongBreak != nil && *req.LongBreak >= 300 && *req.LongBreak <= 2700 {
			pomodoroSettings.LongBreak = *req.LongBreak
		}
		if req.CyclesUntilLong != nil && *req.CyclesUntilLong >= 2 && *req.CyclesUntilLong <= 8 {
			pomodoroSettings.CyclesUntilLong = *req.CyclesUntilLong
		}
		if req.ShowInCycle != nil {
			pomodoroSettings.ShowInCycle = *req.ShowInCycle
		}

		// Update session time if not active
		if !pomodoroSession.Active {
			pomodoroSession.TimeRemaining = pomodoroSettings.WorkDuration
		}

		response := map[string]interface{}{
			"session":  pomodoroSession,
			"settings": pomodoroSettings,
		}
		mutex.Unlock()

		go saveConfig()
		log.Printf("🍅 Pomodoro settings updated: work=%dmin, break=%dmin, long=%dmin, cycles=%d",
			pomodoroSettings.WorkDuration/60, pomodoroSettings.BreakDuration/60,
			pomodoroSettings.LongBreak/60, pomodoroSettings.CyclesUntilLong)

		json.NewEncoder(w).Encode(response)
		return
	}

	jsonError(w, "Method not allowed", http.StatusMethodNotAllowed)
}

File: static/js/pomodoro.js




let pomodoroSession = {
  active: false,
  mode: "work",
  timeRemaining: 25 * 60,
  isPaused: false,
  cyclesCompleted: 0,
};

let pomodoroSettings = {
  workDuration: 25 * 60,
  breakDuration: 5 * 60,
  longBreak: 15 * 60,
  cyclesUntilLong: 4,
  showInCycle: false,
};

let pomodoroPollingInterval = null;
let pomodoroSlidersEditing = false; 
let pomodoroLastSliderChange = 0; 
const POMODORO_EDIT_GRACE_PERIOD = 3000; 





function initPomodoro() {
  loadPomodoroState();
  
  if (pomodoroPollingInterval) clearInterval(pomodoroPollingInterval);
  pomodoroPollingInterval = setInterval(loadPomodoroState, 1000);

  
  setupSliderTracking();
}

function setupSliderTracking() {
  const sliders = [
    "workDuration",
    "breakDuration",
    "longBreakDuration",
    "cycleCount",
  ];
  sliders.forEach((id) => {
    const el = document.getElementById(id);
    if (el) {
      el.addEventListener("focus", () => {
        pomodoroSlidersEditing = true;
      });
      el.addEventListener("blur", () => {
        pomodoroSlidersEditing = false;
      });
      el.addEventListener("input", () => {
        pomodoroLastSliderChange = Date.now();
      });
      el.addEventListener("mousedown", () => {
        pomodoroSlidersEditing = true;
      });
      el.addEventListener("mouseup", () => {
        pomodoroSlidersEditing = false;
      });
      el.addEventListener("touchstart", () => {
        pomodoroSlidersEditing = true;
      });
      el.addEventListener("touchend", () => {
        pomodoroSlidersEditing = false;
      });
    }
  });
}





function loadPomodoroState() {
  authFetch("/api/pomodoro")
    .then((res) => res.json())
    .then((data) => {
      if (data.session) {
        pomodoroSession = data.session;
      }
      if (data.settings) {
        pomodoroSettings = data.settings;
      }
      renderPomodoroUI();
    })
    .catch((err) => {
      if (err.message !== "Unauthorized") {
        console.error("Pomodoro load error:", err);
      }
    });
}

function pomodoroAction(action) {
  authFetch("/api/pomodoro", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ action: action }),
  })
    .then((res) => res.json())
    .then((data) => {
      if (data.session) {
        pomodoroSession = data.session;
      }
      if (data.settings) {
        pomodoroSettings = data.settings;
      }
      renderPomodoroUI();
    })
    .catch((err) => {
      if (err.message !== "Unauthorized") {
        console.error("Pomodoro action error:", err);
      }
    });
}

function savePomodoroSettings() {
  const workMin = parseInt(document.getElementById("workDuration").value);
  const breakMin = parseInt(document.getElementById("breakDuration").value);
  const longMin = parseInt(document.getElementById("longBreakDuration").value);
  const cycles = parseInt(document.getElementById("cycleCount").value);
  const showInCycle = document.getElementById("pomodoroInCycle").checked;

  authFetch("/api/pomodoro", {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      workDuration: workMin * 60,
      breakDuration: breakMin * 60,
      longBreak: longMin * 60,
      cyclesUntilLong: cycles,
      showInCycle: showInCycle,
    }),
  })
    .then((res) => res.json())
    .then((data) => {
      if (data.settings) {
        pomodoroSettings = data.settings;
      }
      if (data.session) {
        pomodoroSession = data.session;
      }
      renderPomodoroUI();
      
    })
    .catch((err) => {
      if (err.message !== "Unauthorized") {
        console.error("Pomodoro settings save error:", err);
      }
    });
}





function startPomodoro() {
  pomodoroAction("start");
}

function pausePomodoro() {
  pomodoroAction("pause");
}

function resumePomodoro() {
  pomodoroAction("resume");
}

function resetPomodoro() {
  pomodoroAction("reset");
}

function skipPhase() {
  pomodoroAction("skip");
}

function togglePomodoroInCycle() {
  savePomodoroSettings();
}





function renderPomodoroUI() {
  
  const minutes = Math.floor(pomodoroSession.timeRemaining / 60);
  const seconds = pomodoroSession.timeRemaining % 60;
  const timeStr = `${String(minutes).padStart(2, "0")}:${String(
    seconds
  ).padStart(2, "0")}`;

  const timeEl = document.getElementById("pomodoroTime");
  if (timeEl) timeEl.textContent = timeStr;

  
  const modeEl = document.getElementById("pomodoroMode");
  if (modeEl) {
    let modeText = "READY";
    modeEl.classList.remove("break", "long-break");

    if (pomodoroSession.active) {
      switch (pomodoroSession.mode) {
        case "work":
          modeText = "FOCUS";
          break;
        case "break":
          modeText = "BREAK";
          modeEl.classList.add("break");
          break;
        case "longBreak":
          modeText = "LONG BREAK";
          modeEl.classList.add("long-break");
          break;
      }
      if (pomodoroSession.isPaused) {
        modeText += " (PAUSED)";
      }
    } else {
      modeText = "READY";
    }
    modeEl.textContent = modeText;
  }

  
  const cyclesEl = document.getElementById("pomodoroCycles");
  if (cyclesEl) {
    cyclesEl.textContent = `Cycle ${pomodoroSession.cyclesCompleted}/${pomodoroSettings.cyclesUntilLong}`;
  }

  
  const startBtn = document.getElementById("pomodoroStartBtn");
  if (startBtn) {
    if (!pomodoroSession.active) {
      startBtn.textContent = "▶ Start";
      startBtn.onclick = startPomodoro;
      startBtn.classList.remove("playing");
    } else if (pomodoroSession.isPaused) {
      startBtn.textContent = "▶ Resume";
      startBtn.onclick = resumePomodoro;
      startBtn.classList.remove("playing");
    } else {
      startBtn.textContent = "⏸ Pause";
      startBtn.onclick = pausePomodoro;
      startBtn.classList.add("playing");
    }
  }

  
  updateSettingsUI();
}

function updateSettingsUI() {
  
  const isWithinGracePeriod =
    Date.now() - pomodoroLastSliderChange < POMODORO_EDIT_GRACE_PERIOD;
  if (pomodoroSlidersEditing || isWithinGracePeriod) {
    
    const showInCycleEl = document.getElementById("pomodoroInCycle");
    if (showInCycleEl) {
      showInCycleEl.checked = pomodoroSettings.showInCycle;
    }
    return;
  }

  const workEl = document.getElementById("workDuration");
  const breakEl = document.getElementById("breakDuration");
  const longEl = document.getElementById("longBreakDuration");
  const cyclesEl = document.getElementById("cycleCount");
  const showInCycleEl = document.getElementById("pomodoroInCycle");

  if (workEl) {
    workEl.value = Math.round(pomodoroSettings.workDuration / 60);
    document.getElementById("workDurationVal").textContent = workEl.value;
  }
  if (breakEl) {
    breakEl.value = Math.round(pomodoroSettings.breakDuration / 60);
    document.getElementById("breakDurationVal").textContent = breakEl.value;
  }
  if (longEl) {
    longEl.value = Math.round(pomodoroSettings.longBreak / 60);
    document.getElementById("longBreakVal").textContent = longEl.value;
  }
  if (cyclesEl) {
    cyclesEl.value = pomodoroSettings.cyclesUntilLong;
    document.getElementById("cyclesVal").textContent = cyclesEl.value;
  }
  if (showInCycleEl) {
    showInCycleEl.checked = pomodoroSettings.showInCycle;
  }
}

function updateDurationDisplay(type, value) {
  switch (type) {
    case "work":
      document.getElementById("workDurationVal").textContent = value;
      break;
    case "break":
      document.getElementById("breakDurationVal").textContent = value;
      break;
    case "longBreak":
      document.getElementById("longBreakVal").textContent = value;
      break;
    case "cycles":
      document.getElementById("cyclesVal").textContent = value;
      break;
  }
}





function cleanupPomodoro() {
  if (pomodoroPollingInterval) {
    clearInterval(pomodoroPollingInterval);
    pomodoroPollingInterval = null;
  }
}
